### PR TITLE
GH-33935: [Go][FlightRPC] Implement Flight SQL extensions

### LINKED ISSUE #33935: [Go][FlightRPC] Implement Flight SQL extensions (transactions, Substrait support)
### Describe the enhancement requested

This was done for C++/Java earlier but Go was not included at the time

### Component(s)

FlightRPC, Go

### PR DESCRIPTION
**Pull request by zeroshade**
### Rationale for this change

Bringing the Go implementation of Flight SQL in line with the C++ and Java implementations.

### What changes are included in this PR?

Implementation of Transactions, Save points and more. This includes adding new functions to the Server interface.

### Are these changes tested?
 Yes, by integration testing.

### Are there any user-facing changes?
This is a breaking change as a few methods are simplified to utilize the Client's allocator rather than needing a separate allocator passed in.

* Closes: #33935

### CODE REVIEW

**Review by lidavidm — APPROVED** [review]
Do we want to add unit tests for the new methods, in addition to the integration test?

**Comment by lidavidm** [comment]
That's another question; I assume something differs in something that isn't being printed

**Comment by zeroshade** [comment]
@lidavidm I figured out the issue, I had the metadata at Schema level, but the C++/java impls have the metadata at the column level. Latest change should fix it, hopefully this one should pass now :)

**Review by lidavidm — COMMENTED** [review]
Hmm, why do they print identically? Or was that two spaces of indentation meant to be the difference?

File: go/arrow/internal/flight_integration/scenario.go [hunk 1]
```diff
@@ -526,17 +526,17 @@ func (m *middlewareScenarioTester) GetFlightInfo(ctx context.Context, desc *flig
...
+// 	// Schema to be returned for mocking the statement/prepared statement
+// 	// results. Must be the same across all languages
+// 	QuerySchema = arrow.NewSchema([]arrow.Field{{
+// 		Name: "id", Type: arrow.PrimitiveTypes.Int64, Nullable: true,
+// 		Metadata: flightsql.NewColumnMetadataBuilder().
+// 			TableName("test").IsAutoIncrement(true).IsCaseSensitive(false).
+// 			TypeName("type_test").SchemaName("schema_test").IsSearchable(true).
+// 			CatalogName("catalog_test").Precision(100).Metadata(),
+// 	}}, nil)
+// )
```
[the remark this answers is not in the document]
Comment (zeroshade) [reply go/arrow/internal/flight_integration/scenario.go]: i've removed it :) it was unncessary

**Comment by zeroshade** [comment]
@lidavidm the two spaces of indentation was supposed to indicate it was metadata for the field, schema level metadata would print out lined up with the top (no indentation) and column-level metadata prints out lined up with the column printout

**Review by lidavidm — APPROVED** [review]

**Pull request merged**